File: android-beans/src/test/java/rocks/frieler/android/beans/scopes/ScopedFactoryBeanDecoratorTest.java
package rocks.frieler.android.beans.scopes;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java8.util.function.Function;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.sameInstance;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static rocks.frieler.android.beans.scopes.ScopedFactoryBeanDecorator.decorate;

@RunWith(MockitoJUnitRunner.class)
public class ScopedFactoryBeanDecoratorTest {
    @SuppressWarnings("unchecked")
    private ScopedFactoryBean<ScopedFactoryBeanDecoratorTest> scopedFactoryBean = mock(ScopedFactoryBean.class);
    private ScopedFactoryBeanDecorator<ScopedFactoryBeanDecoratorTest> decoratedFactoryBean = decorate(scopedFactoryBean);

    @Test
    public void testGetScopeDelegatesToOriginal() {
        when(scopedFactoryBean.getScope()).thenReturn("scope");

        final String scope = decoratedFactoryBean.getScope();

        assertThat(scope, is(scopedFactoryBean.getScope()));
    }

    @Test
    public void testGetBeanTypeDelegatesToOriginal() {
        when(scopedFactoryBean.getBeanType()).thenReturn(ScopedFactoryBeanDecoratorTest.class);

        final Class<?> beanType = decoratedFactoryBean.getBeanType();

        assertThat(beanType, is(equalTo(scopedFactoryBean.getBeanType())));
    }

    @Test
    public void testProduceBeanDelegatesToOriginal() {
        when(scopedFactoryBean.produceBean()).thenReturn(this);

        final ScopedFactoryBeanDecoratorTest bean = decoratedFactoryBean.produceBean();

        assertThat(bean, is(sameInstance(scopedFactoryBean.produceBean())));
    }

    @Test
    public void testProduceBeanAppliesPostProcessingWhenConfigured() {
        when(scopedFactoryBean.produceBean()).thenReturn(this);
        @SuppressWarnings("unchecked") final Function<ScopedFactoryBeanDecoratorTest, ScopedFactoryBeanDecoratorTest> postProcessing = mock(Function.class);
        when(postProcessing.apply(scopedFactoryBean.produceBean())).thenReturn(new ScopedFactoryBeanDecoratorTest());

        final ScopedFactoryBeanDecoratorTest bean = decoratedFactoryBean.withPostProcessing(postProcessing).produceBean();

        verify(postProcessing).apply(scopedFactoryBean.produceBean());
        assertThat(bean, is(sameInstance(postProcessing.apply(scopedFactoryBean.produceBean()))));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/scopes/ScopedFactoryBeanDecoratorTest.kt
package rocks.frieler.android.beans.scopes

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isSameAs
import org.junit.jupiter.api.Test
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import rocks.frieler.android.beans.BeansProvider
import kotlin.reflect.KClass

class ScopedFactoryBeanDecoratorTest {
    private val scopedFactoryBean: ScopedFactoryBean<ScopedFactoryBeanDecoratorTest> = mock()
    private val decoratedFactoryBean = scopedFactoryBean.decorate()

    @Test
    fun `scope delegates to original`() {
        whenever(scopedFactoryBean.scope).thenReturn("scope")

        val scope = decoratedFactoryBean.scope

        assertThat(scope).isEqualTo(scopedFactoryBean.scope)
    }

    @Test
    fun `beanType delegates to original`() {
        whenever(scopedFactoryBean.beanType).thenReturn(ScopedFactoryBeanDecoratorTest::class)

        val beanType: KClass<*> = decoratedFactoryBean.beanType

        assertThat(beanType).isEqualTo(scopedFactoryBean.beanType)
    }

    @Test
    fun `produceBean() delegates to the original`() {
        val dependencies: BeansProvider = mock()
        whenever(scopedFactoryBean.produceBean(dependencies)).thenReturn(this)

        val bean = decoratedFactoryBean.produceBean(dependencies)

        verify(scopedFactoryBean).produceBean(dependencies)
        assertThat(bean).isSameAs(this)
    }

    @Test
    fun `produceBean() applies postProcessing when configured`() {
        val dependencies: BeansProvider = mock()
        val bean = this
        whenever(scopedFactoryBean.produceBean(dependencies)).thenReturn(bean)
        val postProcessing: (ScopedFactoryBeanDecoratorTest) -> ScopedFactoryBeanDecoratorTest = mock()
        val beanAfterPostProcessing = ScopedFactoryBeanDecoratorTest()
        whenever(postProcessing.invoke(bean)).thenReturn(beanAfterPostProcessing)

        val finalBean = decoratedFactoryBean.withPostProcessing(postProcessing).produceBean(dependencies)

        inOrder(scopedFactoryBean, postProcessing) {
            verify(scopedFactoryBean).produceBean(dependencies)
            verify(postProcessing).invoke(bean)
        }
        assertThat(finalBean).isSameAs(beanAfterPostProcessing)
    }
}
